### PR TITLE
[incubator][VC ] add mock testing framework

### PR DESCRIPTION
**Pull request by zhuangqh**
1. make each component testable
2. add test for namespace downward syncer 

Signed-off-by: zhuangqh <zhuangqhc@gmail.com>

### CODE REVIEW

File: incubator/virtualcluster/pkg/syncer/cluster/cluster.go [hunk 1]
```diff
@@ -58,9 +58,6 @@ type Cluster struct {
 	// UID of the corresponding virtual cluster object.
 	vcUID string
 
-	// KubeClientConfig is used to make it easy to get an api server client. Required.
-	KubeClientConfig clientcmd.ClientConfig
```
Comment (zhuangqh) [inline incubator/virtualcluster/pkg/syncer/cluster/cluster.go]: actually no one use it.

File: incubator/virtualcluster/pkg/syncer/cluster/fake_cluster.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	fakeClientset clientset.Interface
+	fakeClient    client.Client
+}
+
+var _ mccontroller.ClusterInterface = &fakeCluster{}
+
+func NewFakeTenantCluster(vc *v1alpha1.Virtualcluster, fakeClientSet clientset.Interface, fakeClient client.Client) (*fakeCluster, error) {
+	cluster := &fakeCluster{
+		key:           conversion.ToClusterKey(vc),
+		vc:            vc,
```
Comment (zhuangqh) [inline incubator/virtualcluster/pkg/syncer/cluster/fake_cluster.go]: static vc object. If there are any change to vc when testing later, please put vc inside a vclister

File: incubator/virtualcluster/pkg/syncer/resources/namespace/dws.go [hunk 3]
```diff
@@ -95,25 +99,24 @@ func (c *controller) reconcileNamespaceCreate(clusterName, targetNamespace, requ
 		return err
 	}
 
-	pNamespace, err := c.namespaceClient.Namespaces().Create(newObj.(*v1.Namespace))
+	_, err = c.namespaceClient.Namespaces().Create(newObj.(*v1.Namespace))
```
Comment (zhuangqh) [inline incubator/virtualcluster/pkg/syncer/resources/namespace/dws.go]: return nil object when create fail, we cannot use the pNamespace

File: incubator/virtualcluster/pkg/syncer/resources/namespace/dws_test.go [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+}
+
+func runDownwardSync(
+	testTenant *v1alpha1.Virtualcluster,
+	existingObjectInSuper []runtime.Object,
+	existingObjectInTenant *v1.Namespace,
+	enqueueObject *v1.Namespace,
+) (actions []core.Action, reconcileError error, err error) {
+	// setup fake tenant cluster
+	tenantClientset := fake.NewSimpleClientset()
```
Comment (zhuangqh) [inline incubator/virtualcluster/pkg/syncer/resources/namespace/dws_test.go]: fake clientset take effects on final actions
fake client take effects on informers.

File: incubator/virtualcluster/pkg/syncer/resources/service/checker.go [hunk 5]
```diff
@@ -117,7 +116,7 @@ func (c *controller) checkServicesOfTenantCluster(clusterName string) {
 		targetNamespace := conversion.ToSuperMasterNamespace(clusterName, vService.Namespace)
 		pService, err := c.serviceLister.Services(targetNamespace).Get(vService.Name)
 		if errors.IsNotFound(err) {
-			if err := c.multiClusterServiceController.RequeueObject(clusterName, &svcList.Items[i], reconciler.AddEvent); err != nil {
```
Comment (zhuangqh) [inline incubator/virtualcluster/pkg/syncer/resources/service/checker.go]: actually no one use it

File: incubator/virtualcluster/pkg/syncer/resources/namespace/dws_test.go [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+			if len(tc.ExpectedCreatedNamespace) != len(actions) {
+				t.Errorf("%s: Expected to create namespace %#v. Actual actions were: %#v", k, tc.ExpectedCreatedNamespace, actions)
+				return
+			}
+			for i, expectedName := range tc.ExpectedCreatedNamespace {
+				action := actions[i]
+				if !action.Matches("create", "namespaces") {
+					t.Errorf("%s: Unexpected action %s", k, action)
+				}
+				createdNS := action.(core.CreateAction).GetObject().(*v1.Namespace)
```
Comment (zhuangqh) [inline incubator/virtualcluster/pkg/syncer/resources/namespace/dws_test.go]: simply check the creation action. The spec checking should be done in the UT for package `conversion`

**Comment by Fei-Guo** [comment]
Nice start.
/lgtm
/approve

**Pull request merged**